Upload iOS peak memory usage metric

huydhn opened the pull request:
This extracts iOS `Memory Peak Physical, kB` metric and uses it as the benchmark `peak_mem_usage(mb)`. I convert the value to mb because it feels friendlier to human looking at the dashboard, but let me know if we need kb granularity.

Also fix a small regex bug when parsing model name like `tinyllama_xnnpack+custom+qe_fp32` where the backend could include `+`

This also fixes the missing key error in https://github.com/pytorch/executorch/actions/runs/11337625751/job/31531760841#step:7:112, it happened because the script failed to parse `tinyllama_xnnpack+custom+qe_fp32` and got back no benchmark records.

huydhn commented:
Testing run https://github.com/pytorch/executorch/actions/runs/11360762990

huydhn commented:
Another thing I notice is that `XCTClockMetric` returns float value in seconds.  And, while the script converts it to ms by x1000, the end value doesn't have the same level of accuracy as its Android counterpart.  This explains @guangy10 comment on https://github.com/pytorch/test-infra/pull/5758#discussion_r1799884428

huydhn commented:
@guangy10 This might answer your question earlier about how people can use the dashboard.  Here is the results from my test run above https://github.com/pytorch/executorch/actions/runs/11360762990, I can see the results on the dashboard, and compare with other branch, i.e. main

https://hud.pytorch.org/benchmark/llms?startTime=Wed%2C%2009%20Oct%202024%2017%3A24%3A45%20GMT&stopTime=Wed%2C%2016%20Oct%202024%2017%3A24%3A45%20GMT&granularity=hour&lBranch=add-memory-metric-ios&lCommit=37a00ca41c98ebbd97e060d11987f00918b67037&rBranch=add-memory-metric-ios&rCommit=37a00ca41c98ebbd97e060d11987f00918b67037&repoName=pytorch%2Fexecutorch&modelName=All%20Models&dtypeName=All%20DType&deviceName=All%20Devices

guangy10 commented:
The app measures two `peak_mem_usage` one for load time including load weights, the other one for forward/execute time. We should display both probably. To me the load time mem usage seems more important especially for large LLMs, if we choose to display one true `peak_mem_usage` it should be the load time one. We need to clearly tell the difference to avoid confusion if we display both. @shoumikhin could explain more.

@huydhn Not a blocker but I'm curious if there is a way to add/display annotation when hovering on the filed name, something like we see in the scuba table? 
![Screenshot 2024-10-16 at 10 36 31 AM](https://github.com/user-attachments/assets/a9982893-749b-416a-a8cf-4450512517fc)

guangy10 commented:
> And, while the script converts it to ms by x1000, the end value doesn't have the same level of accuracy as its Android counterpart.

Why is it? It shouldn't affect the accuracy

huydhn commented:
> > And, while the script converts it to ms by x1000, the end value doesn't have the same level of accuracy as its Android counterpart.
> 
> Why is it? It shouldn't affect the accuracy

Ah, the number is usual something like 0.003 seconds, which is turned into 3 ms.  I guess you want to know why they don't have the same accuracy as those numbers on Android

huydhn commented:
> The app measures two `peak_mem_usage` one for load time including load weights, the other one for forward/execute time. We should display both probably. To me the load time mem usage seems more important especially for large LLMs, if we choose to display one true `peak_mem_usage` it should be the load time one. We need to clearly tell the difference to avoid confusion if we display both. @shoumikhin could explain more.
> 
> @huydhn Not a blocker but I'm curious if there is a way to add/display annotation when hovering on the filed name, something like we see in the scuba table? ![Screenshot 2024-10-16 at 10 36 31 AM](https://private-user-images.githubusercontent.com/42389959/377168243-a9982893-749b-416a-a8cf-4450512517fc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjkxMDA1MjcsIm5iZiI6MTcyOTEwMDIyNywicGF0aCI6Ii80MjM4OTk1OS8zNzcxNjgyNDMtYTk5ODI4OTMtNzQ5Yi00MTZhLWE4Y2YtNDQ1MDUxMjUxN2ZjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDE2VDE3MzcwN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTFhNDk3NWI2YTRkOWI1OTE5YTRlZWE0ZTA1YWE4NWUzYzc1N2VkMjE5M2UwNzEyMzE5ZTM1NWQ3NWQ5MmY5MDUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.otHDxPKbFBPvlypvg_i7Exs4giPY2AwBN6yIH_EZKCU)

huydhn commented:
> @huydhn Not a blocker but I'm curious if there is a way to add/display annotation when hovering on the filed name, something like we see in the scuba table? ![Screenshot 2024-10-16 at 10 36 31 AM](https://private-user-images.githubusercontent.com/42389959/377168243-a9982893-749b-416a-a8cf-4450512517fc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjkxMDA1MjcsIm5iZiI6MTcyOTEwMDIyNywicGF0aCI6Ii80MjM4OTk1OS8zNzcxNjgyNDMtYTk5ODI4OTMtNzQ5Yi00MTZhLWE4Y2YtNDQ1MDUxMjUxN2ZjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDE2VDE3MzcwN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTFhNDk3NWI2YTRkOWI1OTE5YTRlZWE0ZTA1YWE4NWUzYzc1N2VkMjE5M2UwNzEyMzE5ZTM1NWQ3NWQ5MmY5MDUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.otHDxPKbFBPvlypvg_i7Exs4giPY2AwBN6yIH_EZKCU)

This should be doable.  Ping me the description that you have in mind for each of these fields, I could add them into the database as a description field.

guangy10 commented:
> > > And, while the script converts it to ms by x1000, the end value doesn't have the same level of accuracy as its Android counterpart.
> > 
> > 
> > Why is it? It shouldn't affect the accuracy
> 
> Ah, the number is usual something like 0.003 seconds, which is turned into 3 ms. I guess you want to know why they don't have the same accuracy as those numbers on Android

@huydhn Oh, it may be truncated somewhere when reading it out from the test. Because I can see the raw data shows more than 10 decimal places
![Screenshot 2024-10-16 at 11 02 51 AM](https://github.com/user-attachments/assets/0936106e-06b2-4f92-9c72-dcfc7ee31cbf)
The job link where I got the data: https://github.com/pytorch/executorch/actions/runs/11360762990/job/31600157567

guangy10 commented:
> I see. What do you think about showing the bigger number between the twos? It's also easy to have both of them under two different names though, i.e. `peak_mem_usage (load)` and `peak_mem_usage (forward/generate)` for example.
> 
> I think I will go with the latter approach of showing both, then we can take another look at the dashboard to see how they are displayed

@huydhn `peak_mem_usage (load)` and `peak_mem_usage (forward/generate)` look great to me. We can show both and let people decide whether the metric is helpful. One minor concern is that the Android side might not be able to do the same because per @kirklandsign there is no equivalent Android API that can accurately measure the mem usage.

guangy10 commented:
> > @huydhn Not a blocker but I'm curious if there is a way to add/display annotation when hovering on the filed name, something like we see in the scuba table? ![Screenshot 2024-10-16 at 10 36 31 AM](https://private-user-images.githubusercontent.com/42389959/377168243-a9982893-749b-416a-a8cf-4450512517fc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjkxMDA1MjcsIm5iZiI6MTcyOTEwMDIyNywicGF0aCI6Ii80MjM4OTk1OS8zNzcxNjgyNDMtYTk5ODI4OTMtNzQ5Yi00MTZhLWE4Y2YtNDQ1MDUxMjUxN2ZjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDE2VDE3MzcwN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTFhNDk3NWI2YTRkOWI1OTE5YTRlZWE0ZTA1YWE4NWUzYzc1N2VkMjE5M2UwNzEyMzE5ZTM1NWQ3NWQ5MmY5MDUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.otHDxPKbFBPvlypvg_i7Exs4giPY2AwBN6yIH_EZKCU)
> 
> This should be doable. Ping me the description that you have in mind for each of these fields, I could add them into the database as a description field.

Thanks @huydhn! Here are the annotation you can add to the DB.
 - `peak_mem_usage (load)`:"Peak memory usage during the execution of forward() method, which does not include the memory consumed by the application outside"
 - `peak_mem_usage (execute)`: "Peak memory usage during model loading, including loading the checkpoint".
@shoumikhin Does the description sound accurate?

huydhn commented:
> @huydhn ~Do you have any idea why?~ Oh, actually the `upload-benchmark-results` job fails at the end: https://hud.pytorch.org/hud/pytorch/executorch/main/1?per_page=50&name_filter=apple-perf.
> 
> ```
> 2024-10-16T02:13:26.5041722Z     benchmark_results = transform(
> 2024-10-16T02:13:26.5042810Z   File "/home/ec2-user/actions-runner/_work/executorch/executorch/.github/scripts/extract_benchmark_results.py", line 298, in transform
> 2024-10-16T02:13:26.5043807Z     r["deviceInfo"]["device"] = job_name
> 2024-10-16T02:13:26.5044296Z KeyError: 'deviceInfo'
> ```

Yeah, this is the regex parsing bug that I fix as part of this PR.  That's why you see the results showing up on my test branch

huydhn commented:
Ty for the review!  I will add the description and figure out a way to get the higher resolution results from xcresults in separate PRs later

guangy10 commented:
> Ty for the review! I will add the description and figure out a way to get the higher resolution results from xcresults in separate PRs later

Sounds good to me. BTW, on the dashboard side would you mind split the backend out from the model name ([T204741729](https://www.internalfb.com/intern/tasks/?t=204741729))? If non-trivial it's totally okay to deal with it after the release.

guangy10 commented:
> https://hud.pytorch.org/benchmark/llms?startTime=Wed%2C%2009%20Oct%202024%2017%3A24%3A45%20GMT&stopTime=Wed%2C%2016%20Oct%202024%2017%3A24%3A45%20GMT&granularity=hour&lBranch=add-memory-metric-ios&lCommit=37a00ca41c98ebbd97e060d11987f00918b67037&rBranch=add-memory-metric-ios&rCommit=37a00ca41c98ebbd97e060d11987f00918b67037&repoName=pytorch%2Fexecutorch&modelName=All%20Models&dtypeName=All%20DType&deviceName=All%20Devices

@huydhn BTW, how do you generate the link with detailed configs? Such a link it's quite helpful. However, mine doesn't show the configs e.g. branch, commits, etc. in the URL. It's just https://hud.pytorch.org/benchmark/llms?repoName=pytorch%2Fexecutorch

huydhn commented:
Oh you need to use the copy snapshot link button 
![Screenshot 2024-10-16 at 16 03 06](https://github.com/user-attachments/assets/8fb4e762-a2ca-499e-83e9-755963d2275e)

huydhn commented:
My test earlier completes https://github.com/pytorch/executorch/actions/runs/11360762990 but due to the HUD outage earlier, the workflow information is lost.  I think I need to re-run this now.  Just FYI, we have a data gap from about 11AM to 4PM today.

huydhn commented:
The two memory metric are here https://hud.pytorch.org/benchmark/llms?startTime=Thu%2C%2010%20Oct%202024%2001%3A08%3A46%20GMT&stopTime=Thu%2C%2017%20Oct%202024%2001%3A08%3A46%20GMT&granularity=hour&lBranch=add-memory-metric-ios&lCommit=af73f8eae1d58e39f7c17366e3969a5451edc012&rBranch=add-memory-metric-ios&rCommit=af73f8eae1d58e39f7c17366e3969a5451edc012&repoName=pytorch%2Fexecutorch&modelName=All%20Models&dtypeName=All%20DType&deviceName=All%20Devices

huydhn committed:
Upload iOS peak memory usage metric (#6282)

Summary:
This extracts iOS `Memory Peak Physical, kB` metric and uses it as the benchmark `peak_mem_usage(mb)`. I convert the value to mb because it feels friendlier to human looking at the dashboard, but let me know if we need kb granularity.

Also fix a small regex bug when parsing model name like `tinyllama_xnnpack+custom+qe_fp32` where the backend could include `+`

This also fixes the missing key error in https://github.com/pytorch/executorch/actions/runs/11337625751/job/31531760841#step:7:112, it happened because the script failed to parse `tinyllama_xnnpack+custom+qe_fp32` and got back no benchmark records.

Pull Request resolved: https://github.com/pytorch/executorch/pull/6282

Reviewed By: guangy10

Differential Revision: D64453877

Pulled By: huydhn
